server/routerlicious/packages/routerlicious-base/src/utils/sessionHelper.ts

```diff
@@ -134,16 +135,32 @@ async function updateExistingSession(
         isSessionActive: false,
     };
     try {
-        await documentsCollection.upsert(
+        const result = await documentsCollection.findOrCreate(
             {
                 documentId,
+                "session.isSessionAlive": false,
             },
             {
+                createTime: document.createTime,
                 deli: updatedDeli ?? document.deli,
-                scribe: updatedScribe ?? document.scribe,
+                documentId: document.documentId,
                 session: updatedSession,
-            },
-            null);
+                scribe: updatedScribe ?? document.scribe,
+                tenantId: document.tenantId,
+                version: document.version,
+            });
+        // There is no document with isSessionAlive as false. It means this session has been discovered by
+        // another call, and we encoutered race condition with different clients writing truth into the cosmosdb
+        // from different clusters.
+        if (!result.existing) {
+            return getSession(ordererUrl,
```

Comment:
We need to make sure that this doesn't go into an infinite loop

Reply:
Adding an explicit count might be better.

Reply:
Got it. Add the counter. Thanks. :)